### ISSUE TITLE
Youtube notification "delayed"

### ISSUE THREAD

**Issue by JasperTecHK** [andyholmes/gnome-shell-extension-gsconnect#545]
**Describe the bug**

SImply put, Youtube on Android sends the notification about the song playing after the song ends and the next one has started. Thus why I quoted "delayed", since it is not really delayed moreso incorrect response.

**Steps To Reproduce:**

1.Play Youtube on Android device
2.Watch as your song changes and then you get the notification on your desktop that your previous song is playing!

**Expected behavior**
To get notified about what song is playing now/next, not previously.
**Screenshots**

If applicable, add screenshots to help explain your problem. You can drag-and-drop or cut-and-paste images directly into this edit window, to include them in your report.

**Support Log**

Please generate a support log ([Instructions](wiki/Help#generate-support-log)) and paste any messages related to this issue between the two ``` lines below.

```

```

**System Details (please complete the following information):**

 - **GSConnect version: 23
   - **Installed from: Ubuntu Software
 - **GNOME/Shell version: 3:30
 - **Distro/Release:Ubuntu 18.04.2 LTS

**GSConnect environment (if applicable):**

Note 9
KdeConnect v.1.12.7

**Comment by ferdnyc** [comment]
@JasperTecHK 

Hmm. Here's the thing... all GSConnect can do is display the notifications sent to it by KDE Connect. And KDE Connect, in turn, can only mirror the notifications it receives, when it receives them.

Does the YouTube app on your device put up a "playing" notification, while a video is playing? It doesn't, on my device. Normally there are no notifications generated at all, during the course of just using the app. (Since it only plays in the foreground, there isn't a lot of point to a notification. It'd be different if it was a background-player app like, say, Google Music, where I can be using a completely different app while it's playing. But — at least on my phone — the YouTube app doesn't work that way. Perhaps yours is different?)

If you're seeing notifications, even after-the-fact... well, I'm only guessing without more information, but my initial thought is that they may be the activity notifications you can enable in Settings > Notifications in the Android YouTube app. If "Activity on my channel" notifications are enabled, and your viewing history is shared publicly on your channel, then conceivably you'd get a notification whenever a video is added to your Watched history — which probably _is_ only done **after** you watch it.

TL;DR, it sounds like this is all beyond GSConnect's (or even KDE Connect's) control, I'm afraid.

**Comment by andyholmes** [comment]
>TL;DR, it sounds like this is all beyond GSConnect's (or even KDE Connect's) control, I'm afraid.

This sounds like it's probably the case. The only way to know for sure is to generate a support log for that notification ([instructions](https://github.com/andyholmes/gnome-shell-extension-gsconnect/wiki/Help#generate-support-log)).

I'm going to close this for now, since I think it's probably an Android thing, but we can reopen if the log shows GSConnect is handling the notification packet incorrectly.

**Comment by JasperTecHK** [comment]
[gsconnect - Unknown.zip](https://github.com/andyholmes/gnome-shell-extension-gsconnect/files/3254866/gsconnect.-.Unknown.zip)
Here's the log. Sorry it took this long, I can't seem to open it, but that's hopefully caught the delayed notification.

**Comment by ferdnyc** [comment]
Thanks @JasperTecHK , that did come through. I'm reattaching the contents here as a simple `.txt` file for convenience. (The ".eeoo2z" extension on the log file itself was the source of difficulties, I suspect.)

[gsconnect - Unknown.log.txt](https://github.com/andyholmes/gnome-shell-extension-gsconnect/files/3258816/gsconnect.-.Unknown.log.txt)

As you suspected, it isn't very informative, as from the GSConnect side all it shows is that GSConnect is receiving notifications and displaying them.

However, it was informative as to one aspect of the issue — the app those notifications are coming from is not the standard Google YouTube app, but rather [YouTube Vanced](https://vanced.app/), a third-party (and non-Google-Play-distributed) alternative client for YouTube that _does_ do all of the things that I mentioned not being part of the standard YouTube client, like background playback. There's also an XDA Developers [support thread for the app](https://forum.xda-developers.com/android/apps-games/app-youtube-vanced-edition-t3758757).

My suspicion is still that the issue isn't with KDE Connect either. (I just can't imagine a scenario where it would delay notifications before sending them; I won't say it's _impossible_, but it feels unlikely.) Without using the app itself, it feels like this is either a bug in YouTube Vanced, or a "working as designed" issue along the lines I originally theorized, where the notifications it publishes are history additions (occurring at the end of playback), rather than "currently-playing" notices (occurring when playback starts). But there might be more info in that support thread.

**Comment by ferdnyc** [comment]
A support thread [search for the word "notifications"](https://forum.xda-developers.com/search/thread/3758757?query=notification) turns up over 100 results, so clearly they are an oft-discussed aspect of the client!

**Comment by ferdnyc** [comment]
> or a "working as designed" issue along the lines I originally theorized, where the notifications it publishes are history additions (occurring at the end of playback), rather than "currently-playing" notices (occurring when playback starts).

I take that part back, because the notification content clearly identifies them as now-playing notifications, including the presence of playback controls.

The interesting thing is that... well, (a) **three** notifications are coming through for the exact same song, and (b) the timing is, indeed, inconsistent. 

* GSConnect receives them all simultaneously, at 15:36:13.
* The packets have `id` parameters in millisecond-resolution epoch time form
* They include a separate `time` parameter, also in that format (but not the same value as the `id` parameter).

When  we translate the `id`/`time` values into clock time, things get weirder. (The argument to `--date` is a timestamp in _second_ resolution, which is the most precision `date` supports for input. It's just the `id` or `time` value with the last three digits chopped off. The value of the `TZ` variable, I deduced from the wall-clock timestamps in your log.)
```console
$ # First notification, "id" then "time"
$ TZ='America/Los_Angeles' date --date='@1559687771' --iso-8601=seconds
2019-06-04T15:36:11-07:00
$ TZ='America/Los_Angeles' date --date='@1559687453' --iso-8601=seconds 
2019-06-04T15:30:53-07:00

$ # Second notification, same order
$ TZ='America/Los_Angeles' date --date='@1559687772' --iso-8601=seconds 
2019-06-04T15:36:12-07:00
$ TZ='America/Los_Angeles' date --date='@1559687453' --iso-8601=seconds 
2019-06-04T15:30:53-07:00

$ # Third notification
$ TZ='America/Los_Angeles' date --date='@1559687772' --iso-8601=seconds 
2019-06-04T15:36:12-07:00
$ TZ='America/Los_Angeles' date --date='@1559687771' --iso-8601=seconds 
2019-06-04T15:36:11-07:00
```

So you've got three notifications coming through simultaneously at 15:36:13, with ID values indicating they were all captured 1 or 2 seconds earlier, but with internal _timestamps_ indicating that the first two are from 15:30:53, over 5 minutes earlier (presumably at the start of the song).

I'm not immediately sure why this would be happening, but it triggered a memory of [something I'd seen in that support thread](https://forum.xda-developers.com/showpost.php?p=79662872&postcount=10755)
> casting is back working for me too <strike>but not notifications.
> I tried the guidance above from Logan but this phone is on android 6 GPE and there's
no gear wheel icon in microg's app settings.</strike>
> 
> spoke too soon, 3 notifications just popped up after turning the Wi-Fi off and back on again.

**Comment by ferdnyc** [comment]
At this point I think we've learned as much as we can from the GSConnect data. I suspect a logcat from the device itself would be needed, to determine whether it's YouTube Vanced or KDE Connect that's misbehaving here.

**Comment by ferdnyc** [comment]
Another [support thread post](https://forum.xda-developers.com/showpost.php?p=79661381&postcount=10743) with potentially-helpful information:

> 5. why my notifications are not working???????
> clearly the changelog after updating the app says users actions are required!
> first thing to do:
> open device's settings -> apps -> vanced microG
> there you got a gear icon at top or at the bottom there is "in app settings/notification" option tap on it
> then tap on google cloud message.
> under google cloud message you will see 1 registered app which is vanced but the status is disconnected so tap on the 3 dots menu and change the ping to manual then switch it back to automatic as a result of that the status will change to connected and notifications will work!
> (see attached screenshots below)
> YT guide thanks for mrdodojo: https://youtu.be/H9MLAzh--vg
> 
> notes:
> * the 30 may apk got microG settings shortcut under vanced settings because miui doesn't have the options mentioned above.
> * if the app is not registered just logout then login again back into your account.
> * sometimes having delay with notification just disable battery optimization for microG.

(That very last suggestion, in particular, is worth trying.)

The referenced screenshots:
<img src="https://user-images.githubusercontent.com/538020/58998771-8bce2100-87d0-11e9-80a8-533a7e072a11.png" width=300 height=578> <img src="https://user-images.githubusercontent.com/538020/58998776-8ffa3e80-87d0-11e9-90a2-f0199ed18e50.png" width=300 height=578>

**Comment by JasperTecHK** [comment]
Alright, but I don't know how they got a GUI for MicroG. I certainly can't find it as an app with GUI, even though I did install it. I'll also update to the latest version too, I think mine might be a few weeks old. 
Do you want to get the logcat before or after the update?

**Comment by ferdnyc** [comment]
> Alright, but I don't know how they got a GUI for MicroG. I certainly can't find it as an app with GUI, even though I did install it. 

Mmm, if I'm interpreting this right, no GUI for MicroG. This part talks about stuff in the MicroG listing on the Applications list of the main device Settings interface:
> open device's settings -> apps -> vanced microG
> there you got a gear icon at top or at the bottom there is "in app settings/notification" option tap on it
> then tap on google cloud message.

And then later on they say that they added a "MicroG Settings" link _to Vanced Settings_, because not all devices' built-in Settings apps expose the options described above.

As for the battery optimization, that's also somewhere in the device settings. On my old Galaxy S6 it's in Samsung's Device Maintenance app (which is also accessible via Settings > Device maintenance), there's a Battery section that includes an "Unmonitored apps" list. But I suspect newer Android releases have incorporated those features into the core OS Settings.

However it's done, it's definitely worth adding _both_ MicroG and KDE Connect to the list of apps that aren't subject to power management, as there's a big chance that could be causing notification delays.

> I'll also update to the latest version too, I think mine might be a few weeks old.
> I'll swap over to the Magisk version instead. Perhaps a closer connection to the native android account instead of going through microG might do something different.

Entirely possible, especially if only MicroG relies on Google Cloud Messaging for notifications. It's not at all clear to me whether that's actually the case, but if it's possible to get it out of the loop, can't hurt.
